comparison/comparison.go

```diff
@@ -0,0 +1,509 @@
+		}
+		g := report.New(res.LoadTests[0].Label, helper, dp.config.Report)
+		baseReport, err := g.Generate(res.LoadTests[0].Status.StartTime, res.LoadTests[0].Status.StopTime)
+		if err != nil {
+			return results, fmt.Errorf("error while generating report: %w", err)
+		}
+		g = report.New(res.LoadTests[1].Label, helper, dp.config.Report)
+		newReport, err := g.Generate(res.LoadTests[1].Status.StartTime, res.LoadTests[1].Status.StopTime)
+		if err != nil {
+			return results, fmt.Errorf("error while generating report: %w", err)
+		}
+
+		if c.config.Output.GenerateReport {
+			var buf bytes.Buffer
+			opts := report.CompareOpts{
+				GenGraph:     c.config.Output.GenerateGraphs,
+				GraphsPrefix: fmt.Sprintf("%s%d_", res.deploymentID, res.LoadTests[0].loadTestID),
+			}
+			err := report.Compare(&buf, opts, baseReport, newReport)
+			if err != nil {
+				return results, err
+			}
+			res.Report = buf.String()
+		}
+
+		if c.config.Output.UploadDashboard {
+			var dashboardData bytes.Buffer
+			title := fmt.Sprintf("Comparison - %d - %s - %s",
+				res.LoadTests[0].loadTestID, res.LoadTests[0].Config.DBEngine, res.LoadTests[0].Config.Type)
+			if err := report.GenerateDashboard(title, baseReport, newReport, &dashboardData); err != nil {
+				return results, err
+			}
+
+			url, err := t.UploadDashboard(dashboardData.String())
+			if err != nil {
+				return results, err
+			}
+			res.DashboardURL = fmt.Sprintf("http://%s:3000%s", output.MetricsServer.PublicIP, url)
+		}
+
+		results = append(results, res)
+	}
+
+	return results, nil
+}
+
+func provisionBuilds(t *terraform.Terraform, baseBuildURL, newBuildURL string) error {
+	output, err := t.Output()
+	if err != nil {
+		return err
+	}
+	if !output.HasAppServers() {
+		return errors.New("no app servers in this deployment")
+	}
+
+	extAgent, err := ssh.NewAgent()
+	if err != nil {
+		return err
+	}
+	clients := make([]*ssh.Client, len(output.Instances))
+	for i, instance := range output.Instances {
+		client, err := extAgent.NewClient(instance.PublicIP)
+		if err != nil {
+			return fmt.Errorf("error in getting ssh connection %w", err)
+		}
+		defer client.Close()
+		clients[i] = client
+	}
+
+	for _, client := range clients {
+		for _, url := range []string{baseBuildURL, newBuildURL} {
+			filePrefix := "file://"
+			buildFileName := filepath.Base(url)
+			if strings.HasPrefix(url, filePrefix) {
+				// upload build file from local filesystem
+				buildPath := strings.TrimPrefix(url, filePrefix)
+				info, err := os.Stat(buildPath)
+				if err != nil {
+					return err
+				}
+				if !info.Mode().IsRegular() {
+					return fmt.Errorf("build file %s has to be a regular file", buildPath)
+				}
+				if out, err := client.UploadFile(buildPath, "/home/ubuntu/"+buildFileName, false); err != nil {
+					return fmt.Errorf("error uploading build: %w %s", err, out)
+				}
+			} else {
+				// download build file from URL
+				cmd := fmt.Sprintf("wget -O %s %s", buildFileName, url)
```

Comment:
Just confirming- do the machines have wget installed by default?

Reply:
Yes, I pretty much followed what was already happening in the terraform `remote-exec` provisioner.